src/main/java/com/google/cloud/spanner/pgadapter/wireoutput/ErrorResponse.java

```diff
@@ -36,58 +39,83 @@ public class ErrorResponse extends WireOutput {
   private static final byte HINT_FLAG = 'H';
   private static final byte NULL_TERMINATOR = 0;
 
-  private final byte[] severity;
-  private final byte[] errorMessage;
-  private final byte[] errorState;
-  private final byte[] hints;
-
-  public ErrorResponse(DataOutputStream output, PGException pgException) {
-    super(output, calculateLength(pgException));
-    this.errorMessage = pgException.getMessage().getBytes(StandardCharsets.UTF_8);
-    this.errorState = pgException.getSQLState().getBytes();
-    this.severity = pgException.getSeverity().name().getBytes(StandardCharsets.UTF_8);
-    this.hints =
-        Strings.isNullOrEmpty(pgException.getHints())
-            ? null
-            : pgException.getHints().getBytes(StandardCharsets.UTF_8);
+  private final PGException pgException;
+  private final WellKnownClient client;
+
+  public ErrorResponse(ConnectionHandler connection, PGException pgException) {
+    super(
+        connection.getConnectionMetadata().getOutputStream(),
+        calculateLength(pgException, connection.getWellKnownClient()));
+    this.pgException = pgException;
+    this.client = connection.getWellKnownClient();
   }
 
-  static int calculateLength(PGException pgException) {
+  static int calculateLength(PGException pgException, WellKnownClient client) {
     int length =
         HEADER_LENGTH
             + FIELD_IDENTIFIER_LENGTH
-            + pgException.getSeverity().name().getBytes(StandardCharsets.UTF_8).length
+            + getSeverity(pgException).length
             + NULL_TERMINATOR_LENGTH
             + FIELD_IDENTIFIER_LENGTH
-            + pgException.getSQLState().getBytes().length
+            + getSQLState(pgException).length
             + NULL_TERMINATOR_LENGTH
             + FIELD_IDENTIFIER_LENGTH
-            + pgException.getMessage().getBytes(StandardCharsets.UTF_8).length
+            + getMessage(pgException).length
             + NULL_TERMINATOR_LENGTH
             + NULL_TERMINATOR_LENGTH;
-    if (!Strings.isNullOrEmpty(pgException.getHints())) {
-      length +=
-          FIELD_IDENTIFIER_LENGTH
-              + pgException.getHints().getBytes(StandardCharsets.UTF_8).length
-              + NULL_TERMINATOR_LENGTH;
+    byte[] hints = getHints(pgException, client);
+    if (hints.length > 0) {
+      length += FIELD_IDENTIFIER_LENGTH + hints.length + NULL_TERMINATOR_LENGTH;
     }
     return length;
   }
 
+  static byte[] getSeverity(PGException pgException) {
```

Comment:
nit: should getters be moved into the PGException class? Or is this a class that we do not control

Reply:
We do control that class, but these aren't getters, but methods that convert the severity of the `PGException` to the corresponding PG wire-protocol bytes. I've renamed these methods accordingly. (So these methods should stay here, as we don't want to add PG wire-protocol logic to the `PGException` class)